BOMD

tautomer opened the pull request:
<img width="400" alt="image" src="https://github.com/lanl/OpenMS/assets/15900605/d8951e0f-a55a-4ecf-b199-1456029bd1ab">

Total energy is conserved.

<img width="400" alt="image" src="https://github.com/lanl/OpenMS/assets/15900605/cc2a6117-1a5c-48f6-91e3-6f3382127fd4">

Temperature looks ok.

![untitled](https://github.com/lanl/OpenMS/assets/15900605/f1d6b76f-001f-4286-9bd5-646257831478)

Trajectory looks fine.

Some codes are hardcoded and the whole thing is in the status of "barely works". Still need a lot of work.

tautomer committed:
Fix FDTD python example and apply black formatter

tautomer committed:
Merge branch 'lanl:develop' into develop

tautomer committed:
Merge branch 'develop' of https://github.com/lanl/OpenMS into develop

tautomer committed:
functional BOMD

1. Refactor BaseMD and BOMD classes
2. Add an example for BOMD with ML potential
3. Fix bugs in the hippynn interface

tautomer commented:
A few concerns I think we need address.

1. Backends are not working as they should. We have been assuming numpy throughout the dynamics. I also had to convert the predicted tensors to numpy arrays explicitly. To make it even worse, the `Molecule` object inherited from pyscf assumes numpy internally. We have to overload many functions there to make all backends work.
2. Both ES driver and BaseMD have `Molecule`, but they are out of sync. I think that's what you wanted to do with the dummy function `qm.get_data`, which doesn't exist right now. However, we probably don't need to rely on qm.mol at all. We can pass needed values explicitly.
3. Depends on the thermostat, some might not work in the current framework, like Langevin. You can't do a Verlet propagation and then apply a Langevin thermostat. We might want to think about this part.
4. If we were to define a `BaseMD` and a `manyBaseMD`, we would have to have two different but almost identical `BOMD` to inherent them or a conditional inherence. Neither sounds like a good idea, so I add different functions for single and many molecules in the `BaseMD` and overload the functions at runtime. My only concern is that in the future, for almost every single function related to `Molecule`, we need two versions for it. Do you think we should make the code simpler by doing an if or also wrap single-molecule with a list?
5. We are doing `mol.coords = mol.get_coords` for coordinates and only update `mol.coords` in the dynamics, but the internal coordinates of the pyscf object stay the same. I'm not sure if it will cause problems or confusions in the future. Should we do a proper update of positions with [something like this in pyscf](https://pyscf.org/_modules/pyscf/gto/mole.html#MoleBase.atom_coords:~:text=def%20set_geom_(self%2C%20atoms_or_coords%2C%20unit%3DNone%2C%20symmetry%3DNone%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20inplace%3DTrue)%3A)?
6. Added model systems introduced a new dependency of [cqcpy](https://github.com/awhite862/cqcpy). Since we don't have dependency management, it is not handled right now. Should we add it to the cmake workflow or through requirements.txt and setup.py (I can quickly add them)? I was trying to do it through cmake, but it didn't work out.

ares201005 commented:
Xinyang,


  1.  I haven’t finalized the backend, it’s not working at this moment. Switching backend works in a few demo examples. Forget about the backend at this moment, just use numpy to implement.
  2.  A) Molecule object is till needed in QM, to collect all the molecular properties ( charge, atom type, coordinates, etc) in order to pass to any other ES solvers. B) And molecule in qm can be single molecule and many molecule (for example, QED calculations of many molecule coupled to cavity). C) The molecular object can be taken form MD or standalone input, I.e., QM should functionally independently for single-point calculations.

  1.  This Verlet algorithm eventually be reorganized to incorporate any thermostat.
  2.  Maybe the simplest is indeed wrap the molecule a a list, which is universal for single or many molecules. let’s do in this way.
  3.  Pyscf needs to “build” the molecule to get updated coordinates, or explicit set the coordinates as one for your link, such procedure should be handled in pyscf wrapper (TBA).
  4.  This module dependency will be removed, no need to use to in OpenMS.

Let’s discuss in details next Monday.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Xinyang Li ***@***.***>
Sent: Thursday, October 26, 2023 5:24:22 PM
To: lanl/OpenMS ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [lanl/OpenMS] BOMD (PR #12)


A few concerns I think we need address.

  1.  Backends are not working as they should. We have been assuming numpy throughout the dynamics. I also had to convert the predicted tensors to numpy arrays explicitly. To make it even worse, the Molecule object inherited from pyscf assumes numpy internally. We have to overload many functions there to make all backends work.
  2.  Both ES driver and BaseMD have Molecule, but they are out of sync. I think that's what you wanted to do with the dummy function qm.get_data, which doesn't exist right now. However, we probably don't need to rely on qm.mol at all. We can pass needed values explicitly.
  3.  Depends on the thermostat, some might not work in the current framework, like Langevin. You can't do a Verlet propagation and then apply a Langevin thermostat. We might want to think about this part.
  4.  If we were to define a BaseMD and a manyBaseMD, we would have to have two different but almost identical BOMD to inherent them or a conditional inherence. Neither sounds like a good idea, so I add different functions for single and many molecules in the BaseMD and overload the functions at runtime. My only concern is that in the future, for almost every single function related to Molecule, we need two versions for it. Do you think we should make the code simpler by doing an if or also wrap single-molecule with a list?
  5.  We are doing mol.coords = mol.get_coords for coordinates and only update mol.coords in the dynamics, but the internal coordinates of the pyscf object stay the same. I'm not sure if it will cause problems or confusions in the future. Should we do a proper update of positions with something like this in pyscf<https://urldefense.com/v3/__https://pyscf.org/_modules/pyscf/gto/mole.html*MoleBase.atom_coords:*:text=def*20set_geom_(self*2C*20atoms_or_coords*2C*20unit*3DNone*2C*20symmetry*3DNone*2C*0A*20*20*20*20*20*20*20*20*20*20*20*20*20*20*20*20*20*20inplace*3DTrue)*3A__;I34lJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUl!!Bt8fGhp8LhKGRg!GWia2MINukvOEM-mtDHizQy0DnsPSQKV2R8aaV6lCvzZwj1W7FM8qWgsbVPQ2T4VTrixLuDAJ7e6ay03fhPKxQ$>?
  6.  Added model systems introduced a new dependency of cqcpy<https://urldefense.com/v3/__https://github.com/awhite862/cqcpy__;!!Bt8fGhp8LhKGRg!GWia2MINukvOEM-mtDHizQy0DnsPSQKV2R8aaV6lCvzZwj1W7FM8qWgsbVPQ2T4VTrixLuDAJ7e6ay1aYVXSUQ$>. Since we don't have dependency management, it is not handled right now. Should we add it to the cmake workflow or through requirements.txt and setup.py<https://urldefense.com/v3/__http://setup.py__;!!Bt8fGhp8LhKGRg!GWia2MINukvOEM-mtDHizQy0DnsPSQKV2R8aaV6lCvzZwj1W7FM8qWgsbVPQ2T4VTrixLuDAJ7e6ay21W8xGwQ$> (I can quickly add them)? I was trying to do it through cmake, but it didn't work out.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/lanl/OpenMS/pull/12*issuecomment-1782060715__;Iw!!Bt8fGhp8LhKGRg!GWia2MINukvOEM-mtDHizQy0DnsPSQKV2R8aaV6lCvzZwj1W7FM8qWgsbVPQ2T4VTrixLuDAJ7e6ay2U2sFUsg$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ABA34V7ZQ7B3J7XLQF5OGK3YBLWKNAVCNFSM6AAAAAA6R63KEGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOBSGA3DANZRGU__;!!Bt8fGhp8LhKGRg!GWia2MINukvOEM-mtDHizQy0DnsPSQKV2R8aaV6lCvzZwj1W7FM8qWgsbVPQ2T4VTrixLuDAJ7e6ay1wsjHszg$>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

tautomer committed:
Always wrap Molecule objects inside a list

tautomer commented:
> 1. I haven’t finalized the backend, it’s not working at this moment. Switching backend works in a few demo examples. Forget about the backend at this moment, just use numpy to implement.

Yep, that's what I was expecting, just wanted to make sure I we should plan for something right now.

> A) Molecule object is till needed in QM, to collect all the molecular properties ( charge, atom type, coordinates, etc) in order to pass to any other ES solvers. B) And molecule in qm can be single molecule and many molecule (for example, QED calculations of many molecule coupled to cavity). C) The molecular object can be taken form MD or standalone input, I.e., QM should functionally independently for single-point calculations.

I see. Totally forgot SP part. 
I need think about the best way to construct the `get_data` function. The main thing is that `calculate_forces` has to be a function from the "qm" part, so even if we overload MD's `calculate_forces`, the forces are still stored in "qm" not "MD" (since we are modifying "self.mol", unless we change a function such that it takes a "mol" input and modify it in-place. Correct me if I'm wrong.

As a result, I had to do

```python  
            # position update is in MD          
            self.next_position()
            # TODO: wrap this into a function
            for i in range(len(self.mol)):
                qm.mol[i].coords = self.mol[i].coords

           # force calculation happens in qm
           # I didn't overload the function in MD since it won't solve this problem anyway.
            qm.calculate_force()
            # TODO: wrap this into a function
            for i in range(len(self.mol)):
                mol = self.mol[i]
                for j in range(len(mol.states)):
                    mol.forces = qm.mol[i].states[j].forces
```

which cannot be done with one single call of "get_data`. Initially I just did a brutal-force `self.mol = qm.mol` and `qm.mol = self.mol`.


> This Verlet algorithm eventually be reorganized to incorporate any thermostat.

Ok. If it isn't something we should do right now, I will leave it as is.

> Maybe the simplest is indeed wrap the molecule a a list, which is universal for single or many molecules. let’s do in this way.

Done.



> Pyscf needs to “build” the molecule to get updated coordinates, or explicit set the coordinates as one for your link, such procedure should be handled in pyscf wrapper (TBA).

It seems the function I linked ([set_geom_](https://pyscf.org/_modules/pyscf/gto/mole.html#MoleBase.atom_coords)) can update the internal geometry without rebuilding the molecule. However, since the developers named the function like this, I don't think this is the right thing to do. I only want to use it for its ability of handling units internally. Maybe we can mimic this ourselves.



> This module dependency will be removed, no need to use to in OpenMS.

Ok.

ares201005 committed:
Add examples of FDTD and QEDHF
update models
minor update cmake

ares201005 committed:
Example of binding cpp code to python via pybind11

ares201005 committed:
Add boson/photon class which handles
electron-boson interaction related matrix generation

tautomer committed:
Merge branch 'lanl:develop' into develop

tautomer committed:
Add surface hopping codes

1. Initial draft of SH class
2. Add a standalone Runge-Kutta 4 integrator
3. Update Molecule object accordingly

tautomer committed:
Add unit tests for the RK4 propagator

TODO: numerically the function is way less stable than scipy's

tautomer committed:
fix typos
